_transcripts/034-rust-1.52-1.53.md

```diff
@@ -0,0 +1,1020 @@
+guarantees for Rust, even in a published crate, without the caller knowing that
+this has happened. And so as of 1.52, that is just no longer allowed. If you try
+to set that environment variable in a build script, it just gets ignored by
+Cargo. So it's like a— it's a good just, sort of, additional safety mechanism
+for "don't use this", especially not for this purpose.
+
+__Ben__: I will say that on the issue tracker, I do often see a certain jonhoo's
+name associated with many of these discussions about `RUSTC_BOOTSTRAP`. I wonder
+if he's being a naughty boy, using it in his own projects.
+
+__Jon__: I definitely am, but I have asked many people for many permissions
+first.
+
+__Ben__: Ooh. I'm telling.
+
+__Jon__: I think that's all I had for the 1.52, I guess release series, because
+it had two patch releases. But then that— I guess that brings us to 1.53, which
+is, in and of itself, very exciting. Lots of cool stuff in this one.
+
+__Ben__: Yeah, the coolest thing in this one is definitely the `into_iter`
+implementation for arrays. So this has been, again, a very long saga, in this
+case for the sake of backwards compatibility. I think we talked about this
+previously, one of the— did we?
+
+__Jon__: Yeah, we talked about this in, I think 1.50 or 1.51.
+
+__Ben__: Yeah, I'm not sure— that's right, we were talking about the upcoming
+release, because we were— It's an exciting fun hack that allows like, I think
+that it is entirely natural. It's as a use of const generics, where now that
+const generics are here, and they allow arrays to be kind of like, more first
+class than they used to be, in terms of like, being able to implement things for
+them in a way that makes sense. And it's consistent with the rest of the
+language. Now, you can just write `for i in [1, 2, 3]`. Whereas previously, to
+actually iterate over an array, you'd have to either call like, `.iter()`
+`.into_iter()`, or take it by a slice with a ref in front. Now, you can just
+have a by-value iterator over an array, which is pretty cool, which didn't
+happen for a long time, because specifically, people were calling `.iter()` on
+arrays, which, because of autoref, would give you the iterator for slices, which
+yields elements by reference. And you can work around this by calling, you know,
+have your array `.into_iter().cloned()` to get things out by value. And, like,
+if you have, you know, if types are `Copy`, it's not going to be super expensive
+to get them out, but really you should be able to just do this and it's really
+cool that they found a way to make it happen. Do you want to talk about how they
+made it happen, Jon?
+
+Yeah, the— I mean, the hack is is sort of beautifully simple. The basic idea is
+that in the 2015 and 2018 editions, so basically in all the current editions, if
+you have an array and you call `into_iter` on it, then autoref will sort of take
+place. So autoref is this— we're going to call it a compiler feature or language
```

Comment:
autoref? auto-ref? auto ref?

Reply:
I think autoref [is what the compiler uses](https://github.com/rust-lang/rust/blob/dddebf94bccddaa7b8836380c1d90b34553d79d0/compiler/rustc_infer/src/infer/error_reporting/mod.rs#L2420).